types/time.go

```diff
@@ -789,30 +789,217 @@ func isValidSeparator(c byte, prevParts int) bool {
-func splitDateTime(format string) (seps []string, fracStr string) {
-	index := GetFracIndex(format)
-	if index > 0 {
-		fracStr = format[index+1:]
-		format = format[:index]
+var validIdxCombinations = map[int]struct {
+	h int
+	m int
+}{
+	100: {0, 0}, // 23:59:59Z
+	30:  {2, 0}, // 23:59:59+08
+	50:  {4, 2}, // 23:59:59+0800
+	63:  {5, 2}, // 23:59:59+08:00
+	// postgres supports the following additional syntax that deviates from ISO8601, although we won't support it
+	// currently, it will be fairly easy to add in the current parsing framework
+	// 23:59:59Z+08
+	// 23:59:59Z+08:00
+}
+
+// GetTimezone parses the trailing timezone information of a given time string literal. If idx = -1 is returned, it
+// means timezone information not found, otherwise it indicates the index of the starting index of the timezone
+// information. If the timezone contains sign, hour part and/or minute part, it will be returned as is, otherwise an
+// empty string will be returned.
+//
+// Supported syntax:
+//   MySQL compatible: ((?P<tz_sign>[-+])(?P<tz_hour>[0-9]{2}):(?P<tz_minute>[0-9]{2})){0,1}$, see
+//     https://dev.mysql.com/doc/refman/8.0/en/time-zone-support.html and https://dev.mysql.com/doc/refman/8.0/en/datetime.html
+//     the first link specified that timezone information should be in "[H]H:MM, prefixed with a + or -" while the
+//     second link specified that for string literal, "hour values less than than 10, a leading zero is required.".
+//   ISO-8601: Z|((((?P<tz_sign>[-+])(?P<tz_hour>[0-9]{2})(:(?P<tz_minute>[0-9]{2}){0,1}){0,1})|((?P<tz_minute>[0-9]{2}){0,1}){0,1}))$
+//     see https://www.cl.cam.ac.uk/~mgk25/iso-time.html
+func GetTimezone(lit string) (idx int, tzSign, tzHour, tzSep, tzMinute string) {
+	idx, zidx, sidx, spidx := -1, -1, -1, -1
+	z, s, sc := 0, 0, 0
+	l := len(lit)
+	for i := l - 1; 0 <= i; i-- {
+		if lit[i] == 'Z' {
+			zidx = i
+			break
+		}
+		if sidx == -1 && (lit[i] == '-' || lit[i] == '+') {
+			sidx = i
+		}
+		if spidx == -1 && lit[i] == ':' {
+			spidx = i
+		}
+	}
+	// zidx can be -1 (23:59:59+08:00), l-1 (23:59:59Z)
+	// sidx can be -1, l-3, l-5, l-6
+	// scidx can be -1, l-3
+	// hidx can be -1, l-2, l-4, l-5
+	// midx can be -1, l-2
+	// we could enumerate all valid combinations of these values and look it up in a table, see validIdxCombinations
+	if zidx != -1 {
+		z = l - zidx
+	}
+	if sidx != -1 {
+		s = l - sidx
+	}
+	if spidx != -1 {
+		sc = l - spidx
+	}
+	if z != 1 {
+		z = 0
+	}
+	if s != 3 && s != 5 && s != 6 {
+		s = 0
+	}
+	if sc != 3 {
+		sc = 0
+	}
+	k := z*100 + s*10 + sc
+	if v, ok := validIdxCombinations[k]; ok {
+		hidx, midx := l-v.h, l-v.m
+		valid := func(v string) bool {
+			return '0' <= v[0] && v[0] <= '9' && '0' <= v[1] && v[1] <= '9'
+		}
+		if sidx != -1 {
+			tzSign = lit[sidx : sidx+1]
+			idx = sidx
+		}
+		if zidx != -1 {
+			idx = zidx
+		}
+		if sc != 0 {
+			tzSep = lit[spidx : spidx+1]
+		}
+		if v.h != 0 {
+			tzHour = lit[hidx : hidx+2]
+			if !valid(tzHour) {
```

Comment:
I think use strconv.ParseInt and return int64 is better. Of cource, use int64 can also check the range for hour/minutes.